File: spec/conformance/grammar-check.py
#!/usr/bin/env python3
"""Prove ref-id.json's grammar is engine-neutral: replay vectors.parse against
the Python `re` engine using the python-re adaptation, and report agreement."""
import json
import os
import re
import sys

SPEC_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "ref-id.json")


def load_spec():
    with open(SPEC_PATH, "r", encoding="utf-8") as f:
        return json.load(f)


def build_pattern(spec, engine):
    expr = spec["grammar"]["expression"]
    for old, new in spec["grammar"]["adaptations"][engine]["replace"]:
        expr = expr.replace(old, new)
    return expr


def split_pairs(text, separator):
    """Split `text` on `separator`, then split each piece on the first '='."""
    pairs = []
    for piece in text.split(separator):
        if "=" in piece:
            key, value = piece.split("=", 1)
        else:
            key, value = piece, None
        pairs.append([key, value])
    return pairs


def check_vector(spec, m, vec):
    """Return (agree: bool, reason: str) for one vector against a regex match."""
    expect = vec["expect"]
    status = expect.get("status")
    part = expect.get("part")

    if status == "malformed" and part == "grammar":
        if m is None:
            return True, ""
        return False, "expected no match (malformed/grammar) but it matched"

    if m is None:
        return False, "expected a match but got none"

    gd = m.groupdict()

    # version / type / locator, when present in expect
    if "version" in expect:
        raw = gd.get("version")
        version = spec["version"]["default"] if not raw else int(raw)
        if version != expect["version"]:
            return False, f"version {version!r} != expect {expect['version']!r}"

    if "type" in expect:
        if gd.get("type") != expect["type"]:
            return False, f"type {gd.get('type')!r} != expect {expect['type']!r}"

    if "locator" in expect:
        if gd.get("locator") != expect["locator"]:
            return False, f"locator {gd.get('locator')!r} != expect {expect['locator']!r}"

    # qualifiers, when present in expect: split state group on state.separator,
    # each piece on the first '='
    if "qualifiers" in expect:
        state_sep = spec["grammar"]["state"]["separator"]
        state_group = gd.get("state")
        if state_group:
            actual_pairs = split_pairs(state_group, state_sep)
        else:
            actual_pairs = []
        if actual_pairs != expect["qualifiers"]:
            return False, f"qualifiers {actual_pairs!r} != expect {expect['qualifiers']!r}"

    # fragment, when present and not null in expect
    if expect.get("fragment") is not None:
        frag_sep = spec["grammar"]["fragment"]["separator"]
        frag_group = gd.get("fragment")
        if frag_group is None:
            return False, "expected a fragment group but none captured"
        pieces = frag_group.split(frag_sep)
        path = pieces[0]
        refinements = split_pairs(frag_sep.join(pieces[1:]), frag_sep) if len(pieces) > 1 else []
        exp_frag = expect["fragment"]
        if path != exp_frag.get("path"):
            return False, f"fragment.path {path!r} != expect {exp_frag.get('path')!r}"
        if refinements != exp_frag.get("refinements", []):
            return False, f"fragment.refinements {refinements!r} != expect {exp_frag.get('refinements')!r}"

    return True, ""


def main():
    spec = load_spec()
    pattern = build_pattern(spec, "python-re")
    rx = re.compile(pattern)

    vectors = spec["vectors"]["parse"]
    total = len(vectors)
    agreed = 0
    disagreements = []

    for vec in vectors:
        m = rx.match(vec["input"])
        ok, reason = check_vector(spec, m, vec)
        if ok:
            agreed += 1
        else:
            disagreements.append((vec["name"], reason))

    print(f"python-re: {agreed}/{total} agreed")
    for name, reason in disagreements:
        print(f"  DISAGREE: {name} -- {reason}")

    return 0 if not disagreements else 1


if __name__ == "__main__":
    sys.exit(main())

File: spec/conformance/grammar-check.pl
# SPDX-License-Identifier: Apache-2.0
# Grammar-level conformance: the declared expression, unadapted (pcre2 row), must decompose every parse vector as the JSON says.
#!/usr/bin/env perl
# Prove ref-id.json's grammar is engine-neutral: replay vectors.parse against
# Perl's regex engine, using the unadapted expression (pcre2 adaptation is empty).
use strict;
use warnings;
use JSON::PP qw(decode_json);

use FindBin qw($RealBin);
my $spec_path = "$RealBin/../ref-id.json";

open(my $fh, "<:raw", $spec_path) or die "cannot open $spec_path: $!";
local $/;
my $raw = <$fh>;
close($fh);
my $spec = decode_json($raw);

my $expr = $spec->{grammar}{expression};
# pcre2 adaptation's replace list is empty -- expression used unadapted.
for my $pair (@{ $spec->{grammar}{adaptations}{pcre2}{replace} }) {
    my ($old, $new) = @$pair;
    $expr =~ s/\Q$old\E/$new/g;
}

my $rx = qr/$expr/;

sub split_pairs {
    my ($text, $sep) = @_;
    my @pairs;
    for my $piece (split(/\Q$sep\E/, $text, -1)) {
        if ($piece =~ /=/) {
            my ($k, $v) = split(/=/, $piece, 2);
            push @pairs, [$k, $v];
        } else {
            push @pairs, [$piece, undef];
        }
    }
    return \@pairs;
}

sub pairs_equal {
    my ($a, $b) = @_;
    return 0 unless ref($a) eq 'ARRAY' && ref($b) eq 'ARRAY';
    return 0 unless scalar(@$a) == scalar(@$b);
    for my $i (0 .. $#$a) {
        my ($ak, $av) = @{ $a->[$i] };
        my ($bk, $bv) = @{ $b->[$i] };
        return 0 unless defined($ak) == defined($bk) && (!defined($ak) || $ak eq $bk);
        return 0 unless defined($av) == defined($bv) && (!defined($av) || $av eq $bv);
    }
    return 1;
}

my $vectors = $spec->{vectors}{parse};
my $total = scalar(@$vectors);
my $agreed = 0;
my @disagreements;

for my $vec (@$vectors) {
    my $input = $vec->{input};
    my $expect = $vec->{expect};
    my $status = $expect->{status};
    my $part = $expect->{part};

    my ($ok, $reason) = (1, "");

    if (defined($status) && $status eq 'malformed' && defined($part) && $part eq 'grammar') {
        if ($input =~ $rx) {
            ($ok, $reason) = (0, "expected no match (malformed/grammar) but it matched");
        }
    } else {
        if ($input !~ $rx) {
            ($ok, $reason) = (0, "expected a match but got none");
        } else {
            my %g = %+; # named captures

            if (exists $expect->{version}) {
                my $raw_v = $g{version};
                my $version = (defined($raw_v) && $raw_v ne '') ? int($raw_v) : $spec->{version}{default};
                if ($version != $expect->{version}) {
                    ($ok, $reason) = (0, "version $version != expect $expect->{version}");
                }
            }

            if ($ok && exists $expect->{type}) {
                my $t = $g{type};
                $t = '' unless defined $t;
                if ($t ne $expect->{type}) {
                    ($ok, $reason) = (0, "type '$t' != expect '$expect->{type}'");
                }
            }

            if ($ok && exists $expect->{locator}) {
                my $l = $g{locator};
                $l = '' unless defined $l;
                if ($l ne $expect->{locator}) {
                    ($ok, $reason) = (0, "locator '$l' != expect '$expect->{locator}'");
                }
            }

            if ($ok && exists $expect->{qualifiers}) {
                my $sep = $spec->{grammar}{state}{separator};
                my $state_group = $g{state};
                my $actual_pairs = (defined($state_group) && $state_group ne '')
                    ? split_pairs($state_group, $sep)
                    : [];
                unless (pairs_equal($actual_pairs, $expect->{qualifiers})) {
                    ($ok, $reason) = (0, "qualifiers mismatch");
                }
            }

            if ($ok && exists($expect->{fragment}) && defined($expect->{fragment})) {
                my $fsep = $spec->{grammar}{fragment}{separator};
                my $frag_group = $g{fragment};
                if (!defined($frag_group)) {
                    ($ok, $reason) = (0, "expected a fragment group but none captured");
                } else {
                    my @pieces = split(/\Q$fsep\E/, $frag_group, -1);
                    my $path = shift @pieces;
                    my $refinements = @pieces ? split_pairs(join($fsep, @pieces), $fsep) : [];
                    my $exp_frag = $expect->{fragment};
                    if ($path ne $exp_frag->{path}) {
                        ($ok, $reason) = (0, "fragment.path '$path' != expect '$exp_frag->{path}'");
                    } elsif (!pairs_equal($refinements, $exp_frag->{refinements} // [])) {
                        ($ok, $reason) = (0, "fragment.refinements mismatch");
                    }
                }
            }
        }
    }

    if ($ok) {
        $agreed++;
    } else {
        push @disagreements, [$vec->{name}, $reason];
    }
}

print "perl (pcre2-unadapted): $agreed/$total agreed\n";
for my $d (@disagreements) {
    print "  DISAGREE: $d->[0] -- $d->[1]\n";
}

exit(@disagreements ? 1 : 0);
